qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,420 @@
+from functools import cached_property
+from typing import Dict, Tuple
+
+import cirq
+import numpy as np
+from attrs import frozen
+from cirq_ft.algos.arithmetic_gates import LessThanEqualGate, LessThanGate
+from cirq_ft.algos.multi_control_multi_target_pauli import MultiControlPauli
+from cirq_ft.algos.select_swap_qrom import SelectSwapQROM
+from cirq_ft.linalg.lcu_util import preprocess_lcu_coefficients_for_reversible_sampling
+
+from qualtran import Bloq, BloqBuilder, Register, Signature, SoquetT
+from qualtran.bloqs.arithmetic import EqualsAConstant, GreaterThanConstant, ToContiguousIndex
+from qualtran.bloqs.basic_gates import Hadamard, Ry, Toffoli, XGate
+from qualtran.bloqs.on_each import OnEach
+from qualtran.bloqs.swap_network import CSwapApprox
+from qualtran.cirq_interop import CirqGateAsBloq
+
+
+def add_from_bloq_register_flat_qubits(
+    bb: 'BloqBuilder', cirq_bloq: Bloq, **regs: SoquetT
+) -> Tuple[SoquetT, ...]:
+    """Helper function to split / join registers for cirq gates expeciting single 'qubits' register.
+
+    Args:
+        bb: Bloq builder used during decompostion.
+        cirq_bloq: A CirqGateAsBloq wrapped arithmetic gate.
+        regs: bloq registers we wish to use as flat list of qubits for cirq gate.
+
+    Returns:
+        regs: bloq registers appropriately rejoined following split.
+    """
+    flat_regs = []
+    for _, v in regs.items():
+        if v.reg.bitsize == 1:
+            flat_regs.append([v])
+        else:
+            flat_regs.append(bb.split(v))
+    qubits = np.concatenate(flat_regs)
+    qubits = bb.add(cirq_bloq, qubits=qubits)
+    out_soqs = {}
+    start = 0
+    for _, v in regs.items():
+        if v.reg.bitsize == 1:
+            end = start + 1
+            out_soqs[v] = qubits[start:end][0]
+            start += 1
+        else:
+            end = start + v.reg.bitsize
+            out_soqs[v] = bb.join(qubits[start:end])
+            start += v.reg.bitsize
+    return tuple(s for _, s in out_soqs.items())
+
+
+@frozen
+class UniformSuperpositionTHC(Bloq):
+    r"""Prepare uniform superposition state for THC.
+
+    $$
+        |0\rangle^{\otimes 2\log(M+1)} \rightarrow \sum_{\mu\le\nu}^{M} |\mu\rangle|\nu\rangle
+        + \sum_{\mu}^{N/2}|\mu\rangle|\nu=M+1\rangle,
+    $$
+
+    where $M$ is the THC auxiliary dimension, and $N$ is the number of spin orbitals.
+
+    The toffoli complexity of this gate should be 10 * log(M+1) + 2 b_r - 9.
+    Currently it is a good deal larger due to:
+        1. inverting inequality tests should not need more toffolis.
+        2. We are not using phase-gradient gate toffoli cost for Ry rotations
+        3. Small differences in quoted vs implemented comparator costs.
+
+    See: https://github.com/quantumlib/Qualtran/issues/390
+
+    Args:
+        num_mu: THC auxiliary index dimension $M$
+        num_spin_orb: number of spin orbitals $N$
+
+    Registers:
```

Comment:
can you port these to new registers doc syntax

Reply:
Done.